cpp/src/arrow/util/async_generator_test.cc

```diff
@@ -1102,6 +1105,46 @@ TEST(TestAsyncUtil, ReadaheadFailed) {
   ASSERT_TRUE(IsIterationEnd(definitely_last));
 }
 
+class GeneratorIteratorTestFixture : public GeneratorTestFixture {};
+
+TEST_P(GeneratorIteratorTestFixture, Basic) {
+  ASSERT_OK_AND_ASSIGN(
+      auto it, MakeGeneratorIterator(
+                   [this](internal::Executor* executor) -> AsyncGenerator<TestInt> {
+                     return MakeSource({1, 2, 3}, executor);
+                   }));
+
+  ASSERT_OK_AND_ASSIGN(auto actual, it.ToVector());
+  ASSERT_EQ(std::vector<TestInt>({1, 2, 3}), actual);
+}
+
+TEST_P(GeneratorIteratorTestFixture, Error) {
+  ASSERT_OK_AND_ASSIGN(
+      auto it, MakeGeneratorIterator(
+                   [this](internal::Executor* executor) -> AsyncGenerator<TestInt> {
+                     return FailsAt(MakeSource({1, 2, 3}, executor), 1);
+                   }));
+
+  ASSERT_RAISES(Invalid, it.ToVector());
+}
+
+TEST_P(GeneratorIteratorTestFixture, Transferred) {
+  ASSERT_OK_AND_ASSIGN(auto mock_io_executor, internal::ThreadPool::Make(1));
+  ASSERT_OK_AND_ASSIGN(
+      auto it,
+      MakeGeneratorIterator([this, &mock_io_executor](internal::Executor* executor) {
+        auto source = MakeSource({1, 2, 3}, executor);
+        auto to_io = MakeTransferredGenerator(std::move(source), mock_io_executor.get());
+        auto back = MakeTransferredGenerator(std::move(to_io), executor);
+        return back;
+      }));
+  ASSERT_OK_AND_ASSIGN(auto actual, it.ToVector());
```

Comment:
Just to check my understanding, one call here (when IsSlow()) will synchronously sleep on the CPU executor, attempt to transfer to the I/O executor (which is a no-op as the future is already complete), then transfer back to the CPU executor (also a no-op)?

Reply:
N/A, code removed.